File: cpp/questions/minPathSum.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int rs = grid.size();
        int cs = grid[0].size();
        std::vector<int> rv(cs, INT_MAX);
        rv[0] = 0;
        for (int r=0; r<rs; ++r) {
            rv[0] += grid[r][0];
            for (int c=1; c<cs; ++c) {
                rv[c] = std::min(rv[c], rv[c-1]) + grid[r][c];
            }
        }
        return rv[cs-1];
    }
};

File: python/questions/minPathSum.py
from typing import List

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        gr = len(grid)
        gc = len(grid[0])
        rl = [10**18] * gc
        rl[0] = 0
        for r in range(gr):
            rl[0] += grid[r][0]
            for c in range(1, gc):
                rl[c] = min(rl[c], rl[c-1]) + grid[r][c]
        return rl[-1]
